File: python/1386.cinema-seat-allocation.py
class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: List[List[int]]) -> int:

        rows = defaultdict(int)
        for r,s in reservedSeats:
            s -= 2 # shift unused bit and make 0-based

            if s < 0 or s > 7:
                continue

            rows[r] |= 1 << s

        m1  = 0b00001111 # mask 1: 2,3,4,5
        m2  = 0b11110000 # mask 2: 6,7,8,9
        m3  = 0b00111100 # mask 3: 4,5,6,7
 
        # fill free rows optimally with 2 quater person groups
        total = (n - len(rows)) * 2

        for m in rows.values():
            if m1 & m == 0 or m2 & m == 0 or m3 & m == 0:
                total += 1

        return total

        

File: rust/1386.cinema-seat-allocation.rs
use std::collections::HashMap;

impl Solution {
    pub fn max_number_of_families(n: i32, reserved_seats: Vec<Vec<i32>>) -> i32 {
        // . - unused, 1 - set, 0 - unset
        let m1 = 0x0Fu8; // . 0000 1111 .
        let m2 = 0xF0u8; // . 1111 0000 .
        let m3 = 0x3Cu8; // . 0011 1100 .

        let mut rows: HashMap<i32, u8> = HashMap::new();

        for seat in &reserved_seats {
          let i = seat[0];
          let j = seat[1] - 2;

          if j < 0 || j > 7 {
            continue;
          }

          let mut en = rows.entry(i).or_default();
          *en |= 1 << j; 
        }

        // optimally all empy rows can be filled
        // with the two quad seats
        let mut total = (n - rows.len() as i32) * 2;

        for row in rows.values() {
            if row & m1 == 0 || row & m2 == 0 || row & m3 == 0 {
              total += 1;
            }
        }

        total
    }
}